stock_demand_estimate/wizards/stock_demand_estimate_wizard.py

```diff
@@ -0,0 +1,265 @@
+# Copyright 2016 Eficent Business and IT Consulting Services S.L.
+#   (http://www.eficent.com)
+# Copyright 2016 Aleph Objects, Inc. (https://www.alephobjects.com/)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+
+from odoo import api, fields, models, _
+from odoo.osv import expression
+from odoo.addons import decimal_precision as dp
+from odoo.exceptions import UserError, ValidationError
+
+
+class StockDemandEstimateSheet(models.TransientModel):
+    _name = 'stock.demand.estimate.sheet'
+    _description = 'Stock Demand Estimate Sheet'
+
+    date_start = fields.Date(
+        string="Date From",
+        readonly=True,
+    )
+    date_end = fields.Date(
+        string="Date to",
+        readonly=True,
+    )
+    date_range_type_id = fields.Many2one(
+        string='Date Range Type',
+        comodel_name='date.range.type',
+        readonly=True,
+    )
+    location_id = fields.Many2one(
+        comodel_name="stock.location",
+        string="Location",
+        readonly=True,
+    )
+    line_ids = fields.Many2many(
+        string="Estimates",
+        comodel_name='stock.demand.estimate.sheet.line',
+        relation='stock_demand_estimate_line_rel',
+    )
+    product_ids = fields.Many2many(
+        string="Products",
+        comodel_name="product.product",
+    )
+
+    @api.onchange('date_start', 'date_end', 'date_range_type_id',)
+    def _onchange_dates(self):
+        for sheet in self:
+            if not all([sheet.date_start, sheet.date_end,
+                        sheet.date_range_type_id]):
+                return
+            ranges = sheet._get_ranges()
+            if not ranges:
+                raise UserError(_('There is no ranges created.'))
+            estimates = self.env['stock.demand.estimate'].search([
+                ('product_id', 'in', sheet.product_ids.ids),
+                ('date_range_id', 'in', ranges.ids),
+                ('location_id', '=', sheet.location_id.id),
+            ])
+            lines = []
+            for product in sheet.product_ids:
+                for _range in ranges:
+                    estimate = estimates.filtered(
+                        lambda x: (x.date_range_id == _range and
+                                   x.product_id == product)
+                    )
+                    if estimate:
+                        uom_id = estimate[0].product_uom.id
```

Comment:
@mreficent Please prefer use of :
```suggestion
                        from odoo.fields import first
                        uom_id = first(estimate).product_uom.id
```

Reply:
Just curious, what is the benefit of this change?

Reply:
It's not recommended to use indexes on recordset. Benefits of recordsets are kept. See https://github.com/odoo/odoo/blob/12.0/odoo/fields.py#L48